RavenDB-13671 - Restore from S3

Or-Geva opened the pull request.

Commit:
RavenDB-13671 Add A controller for cloud dir and modify the controller '/admin/restore/points' in order to support cloud backup

ml054 committed:
RavenDB-13671 Restore backup from S3

grisha-kotler committed:
RavenDB-13671 - add restore endpoints options from s3

Commit:
RavenDB-13671 - Restore backup from S3 + tests

Commit:
RavenDB-13671 - Restore backup from S3 + tests

Commit:
RavenDB-13671 - Restore Points

Commit:
RavenDB-13671 - Folder Path

grisha-kotler commented on src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreBackupTaskBase.cs:
```diff
@@ -556,9 +642,11 @@ private void RestoreFromSmugglerFile(Action<IOperationProgress> onProgress, Docu
                 SkipRevisionCreation = true
             };
-            var lastPath = Path.Combine(_restoreConfiguration.BackupLocation, smugglerFile);
+            // var lastPath = Path.Combine(_restoreConfiguration.BackupLocation, smugglerFile);
```
Debug code?

Or-Geva replied:
Fixed

grisha-kotler commented on src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreBackupTaskBase.cs:
```diff
@@ -556,9 +642,11 @@ private void RestoreFromSmugglerFile(Action<IOperationProgress> onProgress, Docu
+            var lastPath = GetBackupPath(smugglerFile);
 
-            using (var zip = ZipFile.Open(lastPath, ZipArchiveMode.Read, System.Text.Encoding.UTF8))
+            //using (var zip = ZipFile.Open(lastPath, ZipArchiveMode.Read, System.Text.Encoding.UTF8))
```
same here

Or-Geva replied:
Fixed

Commit:
RavenDB-13671 rebase + PR comments

ppekrol commented on src/Raven.Client/Documents/Operations/Backups/RestoreBackupConfiguration.cs:
```diff
@@ -18,4 +16,19 @@ public class RestoreBackupConfiguration
+            S3Settings = new S3Settings();
+        }
+
+        public S3Settings S3Settings { get; set; }
```
rename property to `Settings`

Or-Geva replied:
Fixed

ppekrol commented on src/Raven.Client/ServerWide/Operations/RestoreBackupOperation.cs:
```diff
@@ -13,40 +14,61 @@ namespace Raven.Client.ServerWide.Operations
         public string NodeTag;
 
-        public RestoreBackupOperation(RestoreBackupConfiguration restoreConfiguration)
+        public RestoreBackupOperation(RestoreBackupConfiguration restoreConfiguration, string nodeTag = null)
```
adding an argument (even nullable) breaks binary compatibility. Also it would be better just to have `RestoreBackupConfigurationBase` here, then you will not need to have multiple ctors. RestoreBackupConfigurationBase should have an abstract method `ToJson` and property `Type`

Or-Geva replied:
Fixed

ppekrol commented on src/Raven.Client/ServerWide/Operations/RestoreBackupOperation.cs:
```diff
@@ -13,40 +14,61 @@ namespace Raven.Client.ServerWide.Operations
             public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, ServerNode node, out string url)
             {
-                url = $"{node.Url}/admin/restore/database";
+                url = $"{node.Url}/admin/restore/database?type={_restoreType.ToString()}";
```
When `RestoreBackupConfigurationBase` will have a `Type` property you do not need this. You can just send JSON, which will be more convenient

grisha-kotler replied:
We need to send the type anyway.
The endpoint admin/restore/database needs to be backward compatible (it's local by default)

ppekrol replied:
No 'Type' in Json -> Local

Or-Geva replied:
Fixed

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs:
```diff
@@ -146,10 +150,7 @@ private async Task AbortMultiUpload(HttpClient client, string url)
-            {
-                Content = new StringContent(xmlString, Encoding.UTF8, "text/plain")
-            };
+            var requestMessage = new HttpRequestMessage(HttpMethods.Post, url) {Content = new StringContent(xmlString, Encoding.UTF8, "text/plain")};
```
This kind of formatting should be avoided.

Or-Geva replied:
Fixed.

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs:
```diff
@@ -206,10 +207,7 @@ private async Task<string> UploadPart(Stream baseStream, HttpClient client, stri
-                    {
-                        {"Content-Length", subStream.Length.ToString(CultureInfo.InvariantCulture)}
-                    }
+                    Headers = {{"Content-Length", subStream.Length.ToString(CultureInfo.InvariantCulture)}}
```
Formatting

Or-Geva replied:
Fixed.

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs:
```diff
@@ -422,14 +420,11 @@ public async Task PutBucket(string awsRegion = null)
                 var now = SystemTime.UtcNow;
 
                 var hasLocationConstraint = AwsRegion != DefaultRegion;
-                var payloadHash = hasLocationConstraint ?
```
Formatting

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs:
```diff
@@ -422,14 +420,11 @@ public async Task PutBucket(string awsRegion = null)
 
                 var requestMessage = new HttpRequestMessage(HttpMethods.Put, url)
                 {
-                    Content = hasLocationConstraint == false ?
```
Formatting

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs:
```diff
@@ -422,14 +420,11 @@ public async Task PutBucket(string awsRegion = null)
                 {
-                    Content = hasLocationConstraint == false ?
-                        null : new StringContent(xmlString, Encoding.UTF8, "text/plain")
+                    Content = hasLocationConstraint == false ? null : new StringContent(xmlString, Encoding.UTF8, "text/plain")
```
Formatting

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs:
```diff
@@ -500,6 +495,103 @@ public async Task DeleteBucket()
+            if (delimiter != null)
+                url += $"&delimiter={delimiter}";
+
+            //url += $"&max-keys={10}";
```
Debug code?

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs:
```diff
@@ -500,6 +495,103 @@ public async Task DeleteBucket()
+            if (continuationToken != null)
+                url += $"&continuation-token={Uri.EscapeDataString(continuationToken)}";
+
+            //var k = $"?list-type=2&prefix={prefix}";
```
Debug code?

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs:
```diff
@@ -500,6 +495,103 @@ public async Task DeleteBucket()
+                throw StorageException.FromResponseMessage(response);
+
+            var responseStream = await response.Content.ReadAsStreamAsync();
+            var listBucketResult = XDocument.Load(responseStream);
```
Is there a JSON endpoint available?

grisha-kotler replied:
No. Only XML.
https://docs.aws.amazon.com/AmazonS3/latest/dev/ListingObjectKeysUsingREST.html

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreBackupTaskBase.cs:
```diff
@@ -36,49 +36,106 @@
+                throw new ArgumentException("Database name can't be null or empty");
+
+            if (ResourceNameValidator.IsValidResourceName(RestoreFromConfiguration.DatabaseName, _serverStore.Configuration.Core.DataDirectory.FullPath, out string errorMessage) == false)
+                throw new BadRequestException(errorMessage);
```
InvalidOperationException would be better, BadRequest is more related to HTTP

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreBackupTaskBase.cs:
```diff
@@ -36,49 +36,106 @@
+                    if (key.Length != 256 / 8)
+                        throw new InvalidOperationException($"The size of the key must be 256 bits, but was {key.Length * 8} bits.");
+
+                    var isEncrypted = string.IsNullOrWhiteSpace(RestoreFromConfiguration.EncryptionKey) == false;
```
We already know that from previous `if` statement

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreBackupTaskBase.cs:
```diff
@@ -36,49 +36,106 @@
+
+            _restoringToDefaultDataDirectory = IsDefaultDataDirectory(RestoreFromConfiguration.DataDirectory, RestoreFromConfiguration.DatabaseName);
+
+            _hasEncryptionKey = string.IsNullOrWhiteSpace(RestoreFromConfiguration.EncryptionKey) == false;
```
I think this is the 3rd time we are checking that. Can we move that earlier and use the `_hasEncryptionKey` ?

Or-Geva replied:
Yes. moved up and removed duplicate.

Commit:
RavenDB-13671 fix PR comments

ml054 committed:
RavenDB-13671 Provide option to download backups from S3

ml054 committed:
RavenDB-13671 Restore backup from S3 - fixing restore from local path

ml054 committed:
RavenDB-13671 Restore backup from S3 - studio part

ml054 committed:
RavenDB-13671 Restore backup from S3 - fixed listing

ml054 commented on src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs:
```diff
@@ -500,6 +505,101 @@ public async Task DeleteBucket()
+        {
+            var url = $"{GetUrl()}/?list-type=2";
+            if (prefix != null)
+                url += $"&prefix={prefix}";
```
check if we need to escape prefix?

Or-Geva replied:
Does not matter, I will add escape prefix just for safety.

ml054 replied:
ok but please check if it still works after you escape.

ml054 commented on src/Raven.Server/Web/RequestHandler.cs:
```diff
@@ -692,5 +694,22 @@ protected void RedirectToLeader()
             HttpContext.Response.Headers.Add("Location", leaderLocation);
         }
+
+        protected PeriodicBackupConnectionType GetBackupConnectionType()
```
can we find better place for this method? handling backup connection type in abstract request handler is IMHO not right place (to board)

Or-Geva replied:
Fixed, moved to inside the controllers.

ml054 commented on src/Raven.Server/Web/Studio/StudioTasksHandler.cs:
```diff
@@ -71,25 +77,65 @@ public async Task FullDataDirectory()
+                        
+                        break;
+                    case PeriodicBackupConnectionType.S3:
+                        var json = context.ReadForMemory(RequestBodyStream(), "studio-tasks/format");
```
I wonder if we should comment this out, as we won't support listing in s3 buckets for now, as we only support cloud?

Or-Geva replied:
@grisha-kotler

Commit:
RavenDB-13671 fix corrupted local backup location and found a bug at test RavenDB-11139.cs that takes wrong last-backup-file.
However, this test has been successfully passing because at RestoreBackupTask line: "if (file.Equals(RestoreFromConfiguration.LastFileNameToRestore, StringComparison.OrdinalIgnoreCase))" we were getting false all the time
when we compared file(actual file name) VS RestoreFromConfiguration.LastFileNameToRestore that holds the full-path of the file.

Commit:
RavenDB-13671 remove debug code

grisha-kotler commented on test/SlowTests/Server/Documents/PeriodicBackup/RavenDB-11139.cs:
```diff
@@ -1147,12 +1147,13 @@ public async Task CreateSnapshotAndIncrementalBackupWithCompareExchangeAndIdenti
                         TransactionMode = TransactionMode.ClusterWide
                     }))
                     {
+                        WaitForUserToContinueTheTest(store2);
```
Please remove this

Or-Geva replied:
Removed

Commit:
RavenDB-13671 fix PR comments

Commit:
RavenDB-13671 fix restore tests

Commit:
RavenDB-13671 fix tests

ml054 committed:
RavenDB-13671 Restore backup from S3 - ability to restore using link

ppekrol committed:
RavenDB-13671 fixed s3 restore point fetching

ppekrol committed:
RavenDB-13671 dirty fix to make restore work

grisha-kotler committed:
RavenDB-13671 - fix compilation after rebase

grisha-kotler committed:
RavenDB-13671 - parse the date time only if we fail to parse the file name

Commit:
RavenDB-13671 - fix restore point

Commit:
RavenDB-13671 add test for local and S3 restore (check for LastFileToRestore)

Commit:
RavenDB-13671 fix Pr comments (ml054)

Commit:
RavenDB-13671 fix RestoreFromS3.restore_s3_settings_tests
